tools/nd_freeze_install

```diff
@@ -0,0 +1,392 @@
+        ]
+        exec_shell("apt install --fix-broken {} {} {} -y".format(*params[:3]))
+        exec_shell(("apt-get install --no-install-recommends {} {} {} -y {}"
+                    ).format(*params),
+                    env=dict(DEBIAN_FRONTEND="noninteractive"))
+
+
+# Test system for minimum version of apt package
+#
+# Parameters
+# ----------
+# version
+#     string : apt version number to test
+#
+# Returns
+# -------
+# boolean : true if system version is >= to the parameter version
+#
+def is_minimum_apt_version(version):
+    result = exec_shell("dpkg -l apt")
+    matches = re.findall(r'ii\s+apt\s+(\S+)', result)
+    if len(matches) > 0:
+        logging.debug("Found apt version {}".format(matches[0]))
+        command = ["dpkg", "--compare-versions", matches[0], "ge", version]
+        try:
+            result = subprocess.check_call(command)
+            if result == 0:
+                return True
+        except subprocess.CalledProcessError:
+            pass
+    return False
+
+
+if __name__ == '__main__':
+
+    # Determine default architecture if none provided by user
+    if (not args.architecture):
+        args.architecture = exec_shell("dpkg --print-architecture").strip("\n")
+
+    # Set apt setting to allow "outdated" repositories.
+    if not os.path.exists("/etc/apt/apt.conf.d/10no--check-valid-until"):
+        with open("/etc/apt/apt.conf.d/10no--check-valid-until", "w") as fp:
+            fp.write('Acquire::Check-Valid-Until "0";')
+
+    # So we later on could make a decision either we need to clean up after
+    # ourselves. Rely on having a Release file, since cannot be just * since
+    # there could be empty partial/ directory
+    apt_release_list = glob.glob("/var/lib/apt/lists/*Release")
+
+    snapshots_sources_file = '/etc/apt/sources.list.d/snapshots.sources.list'
+
+    for p in args.packages:
+        (package, version) = p.split('=')
+
+        repo_info = get_repo_info(package, version)
+        codename = get_package_codename(repo_info, package)
+
+        if codename:
+            logging.info("Found package {}={} in {} {}".format(package,
+                version, codename, repo_info['component']))
+        else:
+            logging.info(("Failed to find codename for package {}={} in "
+                "snapshot '{}'").format(package, version, repo_info['url']))
+            exit(1)
+
+        write_snapshot_source(snapshots_sources_file, repo_info, codename)
+
+    if args.no_updates:
+        logging.info("Packages were not updated")
+    else:
+        # Update debian-archive-keyring first
+        update_packages(["debian-archive-keyring"])
+        run_apt_get_update()
+        update_packages(args.packages)
+
+    if len(apt_release_list) > 0:
+        logging.info("Cleaning up APT lists because originally there were none")
+        apt_release_list = glob.glob("/var/lib/apt/lists/*Release")
+        for file in apt_release_list:
+            os.remove(file)
+
+    exit(0)
```

Comment:
is there any positive effect of having this explicit `exit(0)` at the end here at all?